build-tools/packages/build-cli/src/commands/generate/typetests.ts

```diff
@@ -0,0 +1,377 @@
+		// If nothing is found when using the "import" condition, try the "require" condition. It may be possible to do this
+		// in a single call to resolve.exports, but the documentation is a little unclear. This seems a safe, if inelegant
+		// solution.
+		// conditions: ["default", "types", "require", "node"]
+		const exports = resolve.exports(packageJson, entrypoint, {
+			conditions: ["types"],
+			require: true,
+		});
+		// Only assign typesPath if it wasn't already assigned earlier.
+		typesPath ??=
+			exports === undefined || exports.length === 0
+				? packageJson.types ?? packageJson.typings
+				: exports[0];
+	} catch {
+		// Catch any exceptions here; we'll return undefined instead of throwing them.
+	}
+
+	return typesPath;
+}
+
+/**
+ * Calculates the file path for type validation tests.
+ *
+ * @param pkg - The package whose type tests are being generated.
+ * @param outDir - The output directory for generated tests.
+ * @param outFile - The filename for generated tests.
+ *
+ * @returns The path to write generated files to.
+ */
+function getTypeTestFilePath(pkg: Package, outDir: string, outFile: string): string {
+	return path.join(
+		pkg.directory,
+		outDir,
+		outFile.includes(unscopedPackageNameString)
+			? outFile.replace(
+					unscopedPackageNameString,
+					changeCase.pascalCase(PackageName.getUnscopedName(pkg.name)),
+				)
+			: outFile,
+	);
+}
+
+/**
+ * Extracts type data from a TS source file and creates a map where each key is a type name and the value is its type
+ * data.
+ *
+ * @param file - The source code file containing type data.
+ * @param log - A logger to use.
+ * @returns The mapping between type name and its type data.
+ */
+function typeDataFromFile(file: SourceFile, log: Logger): Map<string, TypeData> {
+	const typeData = new Map<string, TypeData>();
+	const exportedDeclarations = file.getExportedDeclarations();
+
+	for (const declarations of exportedDeclarations.values()) {
+		for (const declaration of declarations) {
+			for (const typeDefinition of getNodeTypeData(declaration)) {
+				const fullName = getFullTypeName(typeDefinition);
+				if (typeData.has(fullName)) {
+					// This system does not properly handle overloads: instead it only keeps the last signature.
+					log.warning(
+						`Skipping overload for ${fullName}; only the last signature will be used.`,
+					);
+				}
+				typeData.set(fullName, typeDefinition);
+			}
+		}
+	}
+	return typeData;
+}
+
+/**
+ * Loads a ts-morph source file from the provided path.
+ *
+ * @param typesPath - The path to the types file to load. This path is expected to be relative to
+ * @returns The loaded source file.
+ */
+export function loadTypesSourceFile(typesPath: string): SourceFile {
+	const project = new Project({
+		skipAddingFilesFromTsConfig: true,
+		compilerOptions: {
+			module: ModuleKind.Node16,
+			moduleResolution: ModuleResolutionKind.Node16,
+		},
+	});
+	project.addSourceFilesAtPaths(`${path.dirname(typesPath)}/**/*.d.*ts`);
```

Comment:
Why aren't we loading the one file? What is being done with the rest?
(I have been advocating that we don't parse into paths like this. Instead work from the exports just like the generated type test file will do.)